File: Data Structures & Algorithms/validate-parentheses/submission-1.js
class Solution {
    /**
     * @param {string} s
     * @return {boolean}
     */
    isValid(s) {
        if (s.length % 2 !== 0) return false;

        const closeToOpen = {
            ")": "(",
            "}": "{",
            "]": "[",
        };

        const open = {
            "(": true,
            "[": true,
            "{": true,
        };

        let stack = [];

        for (let i = 0; i < s.length; i++) {
            const char = s[i];

            if (open[char]) {
                stack.push(char);
                continue;
            }

            if (closeToOpen[char] !== stack.pop()) return false;
        }

        if (stack.length > 0) return false;

        return true;
    }
}

File: Data Structures & Algorithms/combination-target-sum-ii/submission-1.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []
        candidates.sort()

        def dfs(start, remain_sum, stack):
            if remain_sum == 0:
                res.append(stack.copy())
                return
            if remain_sum < 0:
                return
            for j in range(start, len(candidates)):
                if j > start and candidates[j] == candidates[j - 1]:
                    continue

               

                stack.append(candidates[j])
                dfs(j + 1, remain_sum - candidates[j], stack)
                stack.pop()

        dfs(0, target, [])
        return res
